app/src/org/commcare/android/database/user/UserDatabaseUpgrader.java

```diff
@@ -202,6 +214,62 @@ private boolean upgradeSevenEight(SQLiteDatabase db, int oldVersion, int newVers
         }
     }
 
+    /**
+     * Adding an appId field to FormRecords
+     */
+    private boolean upgradeEightNine(SQLiteDatabase db) {
+        db.beginTransaction();
+        try {
+            if (multipleInstalledAppRecords()) {
+                // Cannot migrate FormRecords once this device has already started installing multiple
+                // applications, because there is no way to know which of those apps the existing
+                // FormRecords belong to
+                throw new MigrationException(true);
```

Comment:
@phillipm This gets propagated back to CommCareApplication and caught here: https://github.com/dimagi/commcare-odk/blob/master/app/src/org/commcare/dalvik/application/CommCareApplication.java#L656, which sets the value of dbState in CCApplication. Then when CCHomeActivity tries to dispatch, it will check the dbState and trigger a handled app exit with the appropriate error message if necessary (https://github.com/dimagi/commcare-odk/blob/master/app/src/org/commcare/dalvik/activities/CommCareHomeActivity.java#L1141-L1151 in CCHomeActivity)